server.py

```diff
@@ -27,6 +30,27 @@ def setup_logger():
             logging.getLogger(logger_name).setLevel(100)
 
 
+def validate_environment():
+    versions = {
+        'python': ('python --version', r'([0-9\.]+)', '3.6.1'),
+        'go': ('go version', r'go([0-9\.]+)', '1.11')
+    }
+    try:
+        for exe, params in versions.items():
+            cmd, pattern, min_version = params
+            failed, output = subprocess.getstatusoutput(cmd)
+            if not failed:
+                version = re.search(pattern, output).group(1)
+                if distutils.version.StrictVersion(version) <= distutils.version.StrictVersion(min_version):
+                    logging.error(f'installed {exe} version {version} is below the required minimum version {min_version}')
```

Comment:
while this is my preferred string formatting, please make it consistent with the rest of the project